cluster-autoscaler/proposals/balance_similar.md

```diff
@@ -0,0 +1,175 @@
+function. Instead the balancing will be executed after expansion option is
+chosen by expansion.Strategy and before node group is resized. The high-level
+algorithm will be as follows:
+1. During loop generating expansion options create a map {node group -> set of
+   pods that pass predicates for this node group}. We already calculate that,
+   just need to store it.
+2. Take expansion option chosen by expansion.Strategy and call it E. Let NG be node group
+   chosen by strategy and K be the number of nodes that need to be added. Let P
+   be the set of all pods that will be scheduled thanks to E.
+3. Find all node groups "similar" to NG and call them NGS.
+4. Check if every pod in P passes scheduler predicates for sample node in every
+   node group in NGS (by checking if P is a subset of set we stored in step 1).
+   Remove from NGS any node group on which at least one pod from P can't be
+   scheduled.
+5. Add NG to NGS.
+6. Get current and maximum size of all node groups in NGS. Split K between node
+   groups as described in example above.
+7. Resize all groups in NGS as per result of step 6.
+
+If the user sets the corresponding flag to 'false' we skip step 3,
+resulting in a single element in NGS (this makes step 4 no-op and step 6 trivial).
+
+## Similar node groups
+We will balance size of similar node groups. We want similar groups to consist
+of machine with the same instance type and with the same set of custom labels.
+In particular we define "similar" node groups as having:
+ * The same Capacity for every resource.
+ * Allocatable for each resource within 5% of each other (this number can depend on a
+   few different factors and so it's good to have some minor slack).
+ * "Free" resources (defined as Allocatable minus resources used by daemonsets and
+   kube-proxy) for each resource within 5% of each other (this number can depend on a
+   few different factors and so it's good to have some minor slack).
+ * The same set of labels, except for zone and hostname labels (defined in
+   https://github.com/kubernetes/kube-state-metrics/blob/master/vendor/k8s.io/client-go/pkg/api/unversioned/well_known_labels.go)
+
+---
+
+## Other possible solutions and discussion
+There are other ways to implement the general idea than the proposed solution.
+This section lists other options that were considered and discusses pros and
+cons of each one. Feel free to skip it.
+
+#### [S1]Split selected expansion option
+This is the solution described in "Implementation proposal" section.
+
+Pros:
+ * Simplest solution.
+
+Cons:
+ * If at least a single pending pod uses zone-based scheduling features the
+   whole scale-up will likely go to a single node group.
+ * We add slightly different nodes than those chosen by expansion.Strategy. In
+   particular any zone-based choices made by expansion.Strategy will be
+   discarded.
+ * We operate on single node groups when creating expansion options, so maximum
```

Comment:
The problem is that if we do the capping after using expander.Strategy, than the strategy cannot use the number of nodes as input. That seems like a massive limitation.

Alternatively we could do all the balancing before applying strategy, which is exactly what [S2] is. However, having a scale up that won't fit in a single node group feels like an edge case and I think it's not worth all the extra complexity. Especially given that we will just do another scale up in 10s.